exporter/splunkhecexporter/heartbeat.go

```diff
@@ -0,0 +1,184 @@
+	defaultHBFailedMetricsName = metricsPrefix + "heartbeats_failed"
+)
+
+type heartbeater struct {
+	config     *Config
+	pushLogFn  func(ctx context.Context, ld plog.Logs) error
+	hbRunOnce  sync.Once
+	hbDoneChan chan struct{}
+
+	// Observability
+	heartbeatSuccessTotal *stats.Int64Measure
+	heartbeatErrorTotal   *stats.Int64Measure
+	tagMutators           []tag.Mutator
+}
+
+func getMetricsName(overrides map[string]string, metricName string) string {
+	if name, ok := overrides[metricName]; ok {
+		return name
+	}
+	return metricName
+}
+
+func newHeartbeater(config *Config, buildInfo component.BuildInfo, pushLogFn func(ctx context.Context, ld plog.Logs) error) *heartbeater {
+	interval := config.Heartbeat.Interval
+	if interval == 0 {
+		return nil
+	}
+
+	var heartbeatsSent, heartbeatsFailed *stats.Int64Measure
+	var tagMutators []tag.Mutator
+	if config.Telemetry.Enabled {
+		overrides := config.Telemetry.OverrideMetricsNames
+		extraAttributes := config.Telemetry.ExtraAttributes
+		var tags []tag.Key
+		tagMutators = []tag.Mutator{}
+		for key, val := range extraAttributes {
+			newTag, _ := tag.NewKey(key)
+			tags = append(tags, newTag)
+			tagMutators = append(tagMutators, tag.Insert(newTag, val))
+		}
+
+		heartbeatsSent = stats.Int64(
+			getMetricsName(overrides, defaultHBSentMetricsName),
+			"number of heartbeats sent",
+			stats.UnitDimensionless)
+
+		heartbeatsSentView := &view.View{
+			Name:        heartbeatsSent.Name(),
+			Description: heartbeatsSent.Description(),
+			TagKeys:     tags,
+			Measure:     heartbeatsSent,
+			Aggregation: view.Sum(),
+		}
+
+		heartbeatsFailed = stats.Int64(
+			getMetricsName(overrides, defaultHBFailedMetricsName),
+			"number of heartbeats failed",
+			stats.UnitDimensionless)
+
+		heartbeatsFailedView := &view.View{
+			Name:        heartbeatsFailed.Name(),
+			Description: heartbeatsFailed.Description(),
+			TagKeys:     tags,
+			Measure:     heartbeatsFailed,
+			Aggregation: view.Sum(),
+		}
+
+		if err := view.Register(heartbeatsSentView, heartbeatsFailedView); err != nil {
+			return nil
+		}
+	}
+
+	hbter := &heartbeater{
+		config:                config,
+		pushLogFn:             pushLogFn,
+		hbDoneChan:            make(chan struct{}),
+		heartbeatSuccessTotal: heartbeatsSent,
+		heartbeatErrorTotal:   heartbeatsFailed,
+		tagMutators:           tagMutators,
+	}
+	hbter.initHeartbeat(buildInfo)
+	return hbter
+}
+
+func (h *heartbeater) shutdown() {
+	close(h.hbDoneChan)
+}
+
+func (h *heartbeater) initHeartbeat(buildInfo component.BuildInfo) {
+	interval := h.config.Heartbeat.Interval
+	if interval == 0 {
+		return
+	}
+
+	h.hbRunOnce.Do(func() {
+		heartbeatLog := h.generateHeartbeatLog(buildInfo)
+		go func() {
+			ticker := time.NewTicker(interval)
+			for {
+				select {
+				case <-h.hbDoneChan:
+					return
+				case <-ticker.C:
+					err := h.pushLogFn(context.Background(), heartbeatLog)
+					h.observe(err)
+				}
+			}
+		}()
```

Comment:
I would recommend moving this `newHeartbeater` under the `if config.Telemetry.Enabled` condition. After that, the heartbeater instance becomes much smaller

```go
	type heartbeater struct {
		doneChan chan struct{}
	}
```

Reply:
Yeah we can move these into `newHeartbeater`. But just a FYI, `config.Telemetry.Enabled` sets whether telemetry is enabled for heartbeating while `config.Heartbeat.Interval` sets whether heartbeat is enabled itself.

Reply:
Yes, if we move it to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20426/files#diff-055d06e38a80dbcd3ed64995921ec9969f705ebc199215a95f7547866600f624R104. that part already has both `config.Heartbeat.Interval` and `config.Telemetry.Enabled` conditions checked.